rustls/examples/internal/test_ca.rs

```diff
@@ -0,0 +1,365 @@
+use std::collections::HashMap;
+use std::env;
+use std::fs::{self, File};
+use std::io::Write;
+use std::net::IpAddr;
+use std::path::PathBuf;
+use std::str::FromStr;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::time::Duration;
+
+use rcgen::{
+    BasicConstraints, Certificate, CertificateParams, CertificateRevocationListParams,
+    DistinguishedName, DnType, ExtendedKeyUsagePurpose, Ia5String, IsCa, KeyIdMethod, KeyPair,
+    KeyUsagePurpose, RevocationReason, RevokedCertParams, RsaKeySize, SanType, SerialNumber,
+    SignatureAlgorithm, PKCS_ECDSA_P256_SHA256, PKCS_ECDSA_P384_SHA384, PKCS_ECDSA_P521_SHA512,
+    PKCS_ED25519, PKCS_RSA_SHA256, PKCS_RSA_SHA384, PKCS_RSA_SHA512,
+};
+use time::OffsetDateTime;
+
+fn main() {
+    // Note: these are ordered such that the data dependencies for issuance are satisfied.
+    let roles = [
+        Role::TrustAnchor,
+        Role::Intermediate,
+        Role::EndEntity,
+        Role::Client,
+    ];
+
+    let sig_algs = [
+        // Note: for convenience we use the RSA sigalg digest algorithm to inform the RSA modulus
+        //   size, mapping SHA256 to RSA 2048, SHA384 to RSA 3072, and SHA512 to RSA 4096.
+        &PKCS_RSA_SHA256,
+        &PKCS_RSA_SHA384,
+        &PKCS_RSA_SHA512,
+        &PKCS_ECDSA_P256_SHA256,
+        &PKCS_ECDSA_P384_SHA384,
+        &PKCS_ECDSA_P521_SHA512,
+        &PKCS_ED25519,
+    ];
+
+    let mut certified_keys = HashMap::with_capacity(roles.len() * sig_algs.len());
+    for role in roles {
```

Comment:
If each `(role, alg)` test is pretty much independent, I don't think we need three loops here? And if that's true, do we still need the hash map?

Reply:
I pulled everything into one loop.

> And if that's true, do we still need the hash map?

It gets fairly awkward without it. We can't strictly look at the last iteration to get the issuer for the current iteration like we do in some other "build a chain" type loops because when we get to the client cert role the last iteration will have been for an end entity cert. We could save the last issuer in separate var only updated for certain roles, but then we also have to do the same for the trust anchor to be able to write a chain file for the EE or client cert role.

In the end since this is a utility only run to generate test data it felt like the approach using the hashmap had the best clarity overall.